File: csharp/Betfair.ESAClient/Betfair.ESAClient/Protocol/ChangeMessage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Protocol
{
    /// <summary>
    /// Adapted version of a change message.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public class ChangeMessage <T>
    {
        private readonly DateTime _arrivalTime;

        public ChangeMessage()
        {
            _arrivalTime = DateTime.UtcNow;
        }

        public long? Pt { get; set; }

        public int Id { get; set; }

        public string Clk { get; set; }

        public string InitialClk { get; set; }

        public long? HeartbeatMs { get; set; }

        public long? ConflateMs { get; set; }

        public List<T> Items { get; set; }

        public SegmentType SegmentType { get; set; }

        public ChangeType ChangeType { get; set; }

        /// <summary>
        /// Start of new subscription (not resubscription)
        /// </summary>
        public bool IsStartOfNewSubscription
        {
            get
            {
                return ChangeType == ChangeType.SUB_IMAGE && 
                    (SegmentType == SegmentType.NONE || SegmentType == SegmentType.SEG_START);
            }
        }

        /// <summary>
        /// Start of subscription / resubscription
        /// </summary>
        public bool IsStartOfRecovery
        {
            get
            {
                return (ChangeType == ChangeType.SUB_IMAGE || ChangeType == ChangeType.RESUB_DELTA) &&
                    (SegmentType == SegmentType.NONE || SegmentType == SegmentType.SEG_START);
            }
        }

        /// <summary>
        /// End of subscription / resubscription
        /// </summary>
        public bool IsEndOfRecovery
        {
            get
            {
                return (ChangeType == ChangeType.SUB_IMAGE || ChangeType == ChangeType.RESUB_DELTA) &&
                    (SegmentType == SegmentType.NONE || SegmentType == SegmentType.SEG_END);
            }
        }

        public DateTime ArrivalTime
        {
            get
            {
                return _arrivalTime;
            }
        }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/protocol/ChangeMessage.java
package com.betfair.esa.client.protocol;

import java.util.List;

/** Created by mulveyj on 07/07/2016. */
public class ChangeMessage<T> {
    private long arrivalTime;
    private long publishTime;
    private int id;
    private String clk;
    private String initialClk;
    private Long heartbeatMs;
    private Long conflateMs;
    private List<T> items;
    private SegmentType segmentType;
    private ChangeType changeType;

    public ChangeMessage() {
        arrivalTime = System.currentTimeMillis();
    }

    /**
     * Start of new subscription (not resubscription)
     *
     * @return
     */
    public boolean isStartOfNewSubscription() {
        return changeType == ChangeType.SUB_IMAGE
                && (segmentType == SegmentType.NONE || segmentType == SegmentType.SEG_START);
    }

    /**
     * Start of subscription / resubscription
     *
     * @return
     */
    public boolean isStartOfRecovery() {
        return (changeType == ChangeType.SUB_IMAGE || changeType == ChangeType.RESUB_DELTA)
                && (segmentType == SegmentType.NONE || segmentType == SegmentType.SEG_START);
    }

    /**
     * End of subscription / resubscription
     *
     * @return
     */
    public boolean isEndOfRecovery() {
        return (changeType == ChangeType.SUB_IMAGE || changeType == ChangeType.RESUB_DELTA)
                && (segmentType == SegmentType.NONE || segmentType == SegmentType.SEG_END);
    }

    public ChangeType getChangeType() {
        return changeType;
    }

    public void setChangeType(ChangeType changeType) {
        this.changeType = changeType;
    }

    public long getArrivalTime() {
        return arrivalTime;
    }

    public void setArrivalTime(long arrivalTime) {
        this.arrivalTime = arrivalTime;
    }

    public long getPublishTime() {
        return publishTime;
    }

    public void setPublishTime(long publishTime) {
        this.publishTime = publishTime;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getClk() {
        return clk;
    }

    public void setClk(String clk) {
        this.clk = clk;
    }

    public String getInitialClk() {
        return initialClk;
    }

    public void setInitialClk(String initialClk) {
        this.initialClk = initialClk;
    }

    public Long getHeartbeatMs() {
        return heartbeatMs;
    }

    public void setHeartbeatMs(Long heartbeatMs) {
        this.heartbeatMs = heartbeatMs;
    }

    public Long getConflateMs() {
        return conflateMs;
    }

    public void setConflateMs(Long conflateMs) {
        this.conflateMs = conflateMs;
    }

    public List<T> getItems() {
        return items;
    }

    public void setItems(List<T> items) {
        this.items = items;
    }

    public SegmentType getSegmentType() {
        return segmentType;
    }

    public void setSegmentType(SegmentType segmentType) {
        this.segmentType = segmentType;
    }
}
